### ISSUE TITLE
v1.10.0 多id的数据集，想选择某一列id做求交和后续训练，求交结果是空的

### ISSUE THREAD

**Issue by MDGBDGMG** [FederatedAI/FATE#4626]
**Describe the bug**
v1.10.0 多id的数据集，想选择某一列id做求交和后续训练，求交结果是空的

**To Reproduce**
1. 数据集说明
party A上传数据集
{
    "file": "/data/projects/fate/examples/data/breast_hetero_guest_repeated_id.csv",
    "table_name": "breast_hetero_guest_repeated_id",
    "namespace": "breast_hetero_guest_repeated_id",
    "head": 1,
    "partition": 1,
    "with_meta": true,
    "meta":
    {
        "delimiter": ",",
        "with_label": true,
        "label_name": "y",
        "input_format": "dense",
        "data_type": "float64",
        "with_match_id": true,
        "id_list": ["id"]
    }
}

partyB 上传数据集
{
    "file": "/data/projects/fate/examples/data/breast_hetero_host.csv",
    "table_name": "breast_hetero_host",
    "namespace": "breast_hetero_host",
    "head": 1,
    "partition": 1
}

<img width="1480" alt="8ea5bca4506ba29645a795b19ee131a" src="https://user-images.githubusercontent.com/45615979/219258105-ddb581b9-83d9-42bd-862d-a96d429f59cf.png">

左边partyA是多id的数据集，想用左边PartyA的id列和右边的id列做求交和训练

3. 执行执行作业
conf文件
{
        "job_parameters":{
                "common":{
                        "job_type":"train",
                        "model_version":"202302071451270667060",
                        "pulsar_run":{},
                        "auto_retries":0,
                        "computing_engine":"EGGROLL",
                        "model_id":"arbiter-1639998474#guest-1639995892#host-1639998474#model",
                        "task_parallelism":1,
                        "rabbitmq_run":{},
                        "engines_address":{},
                        "computing_partitions":4,
                        "federated_status_collect_type":"PUSH",
                        "inheritance_info":{},
                        "auto_retry_delay":1,
                        "adaptation_parameters":{
                                "request_task_cores":4,
                                "task_cores_per_node":4,
                                "task_nodes":1,
                                "if_initiator_baseline":true,
                                "task_memory_per_node":0
                        },
                        "eggroll_run":{},
                        "spark_run":{},
                        "federated_mode":"MULTIPLE"
                }
        },
        "component_parameters":{
                "role":{
                        "host":{
                                "0":{
                                        "task_cores":1,
                                        "reader_0":{
                                                "table":{
                                                        "name":"breast_hetero_host",
                                                        "namespace":"breast_hetero_host"
                                                }
                                        },
                                        "data_transform_0":{
                                                "with_label":false,
"match_id_name": "id"
                                        },
                                        "task_parallelism":1
                                }
                        },
                        "guest":{
                                "0":{
                                        "task_cores":2,
                                        "reader_0":{
                                                "table":{
                                                        "name":"breast_hetero_guest_repeated_id",
                                                        "namespace":"breast_hetero_guest_repeated_id"
                                                }
                                        },
                                        "data_transform_0":{
                                                "label_type":"int",
                                                "with_label":true,
                                                "with_match_id":true,
                                                "match_id_name": "id",
                                                "match_id_index": 1
                                        },
                                        "task_parallelism":2
                                }
                        }
                },
                "common":{
                        "hetero_data_split_0":{
                                "validate_size":0.2,
                                "shuffle":false,
                                "train_size":0.8
                        },
                        "hetero_lr_0":{
                                "batch_size":-1,
                                "penalty":"L2",
                                "early_stop":"diff",
                                "early_stopping_rounds":3,
                                "validation_freqs":1,
                                "cv_param":{
                                        "random_seed":103,
                                        "n_splits":5,
                                        "need_cv":false,
                                        "shuffle":false
                                },
                                "tol":0.0001,
                                "optimizer":"adam",
                                "alpha":0.01,
                                "init_param":{
                                        "init_method":"zeros"
                                },
                                "sqn_param":{
                                        "random_seed":null,
                                        "sample_size":5000,
                                        "update_interval_L":3,
                                        "memory_M":5
                                },
                                "metrics":[],
                                "learning_rate":0.15,
                                "max_iter":3,
                                "use_first_metric_only":false
                        },
                        "hetero_feature_binning_0":{
                                "bin_names":null,
                                "bin_num":5,
                                "method":"bucket",
                                "category_names":null,
                                "error":0.001,
                                "transform_param":{
                                        "transform_type":"bin_num",
                                        "transform_cols":-1,
                                        "transform_names":null
                                },
                                "category_indexes":null,
                                "optimal_binning_param":{
                                        "metric_method":"iv",
                                        "init_bin_nums":100,
                                        "mixture":true,
                                        "init_bucket_method":"quantile",
                                        "max_bin_pct":0.8,
                                        "min_bin_pct":0.05
                                },
                                "compress_thres":10000,
                                "local_only":false,
                                "head_size":10000,
                                "adjustment_factor":0.5,
                                "bin_indexes":-1
                        },
                        "intersection_0":{
                                "intersect_method":"ecdh"
                        },
                        "data_transform_0":{
                                "output_format":"dense",
                                "with_match_id":true
                        },
                        "evaluation_0":{
                                "eval_type":"binary"
                        }
                }
        },
        "dsl_version":2,
        "role":{
                "arbiter":[
                        1639998474
                ],
                "host":[
                        1639998474
                ],
                "guest":[
                        1639995892
                ]
        },
        "conf_path":"heterolr.conf-ecdh",
        "initiator":{
                "role":"guest",
                "party_id":1639995892
        },
        "dsl_path":"dsl.conf"
}

dsl文件：
{
        "components":{
                "hetero_data_split_0":{
                        "output":{
                                "data":[
                                        "train_data",
                                        "validate_data",
                                        "test_data"
                                ]
                        },
                        "input":{
                                "data":{
                                        "data":[
                                                "hetero_feature_binning_0.data"
                                        ]
                                }
                        },
                        "module":"HeteroDataSplit"
                },
                "hetero_lr_0":{
                        "output":{
                                "data":[
                                        "data"
                                ],
                                "model":[
                                        "model"
                                ]
                        },
                        "input":{
                                "data":{
                                        "validate_data":[
                                                "hetero_data_split_0.validate_data"
                                        ],
                                        "train_data":[
                                                "hetero_data_split_0.train_data"
                                        ]
                                }
                        },
                        "module":"HeteroLR"
                },
                "hetero_feature_binning_0":{
                        "output":{
                                "data":[
                                        "data"
                                ],
                                "model":[
                                        "model"
                                ]
                        },
                        "input":{
                                "data":{
                                        "data":[
                                                "intersection_0.data"
                                        ]
                                }
                        },
                        "module":"HeteroFeatureBinning"
                },
                "hetero_lr_1":{
                        "output":{
                                "data":[
                                        "data"
                                ],
                                "model":[
                                        "model"
                                ]
                        },
                        "input":{
                                "data":{
                                        "test_data":[
                                                "hetero_data_split_0.validate_data"
                                        ]
                                },
                                "model":[
                                        "hetero_lr_0.model"
                                ]
                        },
                        "module":"HeteroLR"
                },
                "reader_0":{
                        "output":{
                                "data":[
                                        "data"
                                ]
                        },
                        "module":"Reader"
                },
                "intersection_0":{
                        "output":{
                                "data":[
                                        "data"
                                ]
                        },
                        "input":{
                                "data":{
                                        "data":[
                                                "data_transform_0.data"
                                        ]
                                }
                        },
                        "module":"Intersection"
                },
                "data_transform_0":{
                        "output":{
                                "data":[
                                        "data"
                                ],
                                "model":[
                                        "model"
                                ]
                        },
                        "input":{
                                "data":{
                                        "data":[
                                                "reader_0.data"
                                        ]
                                }
                        },
                        "module":"DataTransform"
                },
                "evaluation_0":{
                        "output":{
                                "data":[
                                        "data"
                                ]
                        },
                        "input":{
                                "data":{
                                        "data":[
                                                "hetero_lr_0.data",
                                                "hetero_lr_1.data"
                                        ]
                                }
                        },
                        "module":"Evaluation"
                }
        }
}

5. 结果
![image](https://user-images.githubusercontent.com/45615979/219258720-0233bbb1-1734-44fb-a2bf-ab3e370bad89.png)

求交模块成功，但是交集为空
![image](https://user-images.githubusercontent.com/45615979/219258609-fe4e0d61-67a0-4051-a272-e749e581b779.png)

binning模块因为输入数据集为空而失败
![image](https://user-images.githubusercontent.com/45615979/219258665-9f0e53b1-c91e-43bb-a555-6a7b218a5f47.png)

**Expected behavior**

**Screenshots**

**Desktop (please complete the following information):**
 - OS:   centos7.9
 - Browser   chrome 
 - Version  v1.10.0

**Smartphone (please complete the following information):**

**Additional context**

**Comment by nemirorox** [comment]
Hi, when one party's data has match id, the other's doesn't, intersection will fail due to mismatched sid. To fix this, I would suggest that configure `extend_sid`=True when uploading host's data, and also set `with_match_id`=True in DataTransform module for host. For demo using data with match-id, please refer [here.](https://github.com/FederatedAI/FATE/blob/master/doc/tutorial/pipeline/pipeline_tutorial_match_id.ipynb)

**Comment by MDGBDGMG** [comment]
按照您给出的方法，确实成功了。但是还有两个问题：1.单id数据集，想和多id数据集做求交，必须加上extend_sid=True吗？sid不可以写入id_list吗？ 2.多id（比方说id1和id2）的数据集，在做训练时，不作为id列的那列数据（选用id1作为match_id，id2不做处理），是做为特征参与训练吗？

**Comment by mgqa34** [comment]
> 按照您给出的方法，确实成功了。但是还有两个问题：1.单id数据集，想和多id数据集做求交，必须加上extend_sid=True吗？sid不可以写入id_list吗？ 2.多id（比方说id1和id2）的数据集，在做训练时，不作为id列的那列数据（选用id1作为match_id，id2不做处理），是做为特征参与训练吗？

A1: 需要先理解匹配ID和样本ID的区别，extend_sid相当于加了1列样本ID。FATE要求要么双方数据都是匹配ID和样本ID做分离的模式，要么两方都不做分离。匹配ID可以理解为就是psi过程中可以做关联的key，而样本ID简单理解其实就是第几行，
A2: upload数据的时候，meta里面的id_list参数给出哪些列是作为候选id列，在实际建模的时候，match_id_name指定使用了哪一列作为当前任务的key，然后id_list的其他列在当前的job里面会被丢弃。

**Comment by MDGBDGMG** [comment]
明白了，多谢指点
